lib/segment/src/vector_storage/memmap_vector_storage.rs

```diff
@@ -85,31 +89,44 @@ impl VectorStorage for MemmapVectorStorage {
         stopped: &AtomicBool,
     ) -> OperationResult<Range<PointOffsetType>> {
         let dim = self.vector_dim();
-
         let start_index = self.mmap_store.as_ref().unwrap().num_vectors as PointOffsetType;
         let mut end_index = start_index;
-
-        self.mmap_store = None;
-
-        {
-            let mut file = OpenOptions::new()
-                .read(false)
-                .write(false)
-                .append(true)
-                .create(false)
-                .open(&self.vectors_path)?;
-
-            for id in other_ids {
-                check_process_stopped(stopped)?;
-                let vector = other.get_vector(id);
-                let raw_bites = vf_to_u8(vector);
-                file.write_all(raw_bites)?;
-                end_index += 1;
+        self.mmap_store.take();
+
+        // Extend vectors file, write other vectors into it
+        let mut vectors_file = open_append(&self.vectors_path)?;
+        let mut deleted_ids = vec![];
+        for id in other_ids {
+            check_process_stopped(stopped)?;
+            let vector = other.get_vector(id);
+            let raw_bites = vf_to_u8(vector);
+            vectors_file.write_all(raw_bites)?;
+            end_index += 1;
+
+            // Remember deleted IDs so we can flush these later
+            if other.is_deleted(id) {
+                deleted_ids.push((start_index + id) as PointOffsetType);
             }
-
-            file.flush()?;
         }
-        self.mmap_store = Some(MmapVectors::open(&self.vectors_path, dim)?);
+        vectors_file.flush()?;
+        drop(vectors_file);
+
+        // Load store with updated files
+        self.mmap_store.replace(MmapVectors::open(
+            &self.vectors_path,
+            &self.deleted_path,
+            dim,
+        )?);
+
+        // Flush deleted flags into store
+        // We must do that in the updated store, and cannot do it in the previous loop. That is
+        // because the file backing delete storage must be resized, and for that we'd need to know
+        // the exact number of vectors beforehand. When opening the store it is done automatically.
+        let store = self.mmap_store.as_mut().unwrap();
+        for id in deleted_ids {
```

Comment:
When updating mmap storage from another, we first insert new vectors in a loop, then we process all deletions in a second loop. I find this ugly, but I don't see a way to do this more elegantly. Is it good enough?

Reply:
We discussed this in a call; this is okay.